cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -689,21 +821,40 @@ auto build_fine_clusters(const handle_t& handle,
                    "Number of fine clusters must be non-zero for a non-empty mesocluster");
     }
 
-    utils::copy_selected(
-      mesocluster_sizes[i], dim, dataset_mptr, mc_trainset_ids, dim, mc_trainset, dim, stream);
+    utils::copy_selected((IdxT)mesocluster_sizes[i],
```

Comment:
This one is used all over the place in both new ivf-* implementations and it uses the `mapping` type to transform between input and output along the way (e.g. case uint8_t to float and then divide by 256).

Reply:
> However, in case of ann_kmeans_balanced, I think we'd better use raft::matrix::copy_rows, because we copy indices here rather than values, and logically it has nothing to do with the ann mapping of the value types.

What do you mean "we copy indices"? The two calls of `copy_selected` in `ann_kmeans_balanced` copy parts of the dataset and the norm respectively. While the norm is `float` and can easily make use of `matrix::copyRows`, the dataset can be `int8`/`uint8` and definitely needs the mapping.

Reply:
But maybe we could adapt `copyRows` to accept an iterator, and then branch inside `matrix::detail::copyRows` to select the most efficient implementation based on whether the iterator is a simple pointer or not.

Reply:
> What do you mean "we copy indices"?

Oh, sorry, I looked the wrong way. Indeed you're right, we do copy values here and do make use of the mapping `T`->`float`. Which, strictly speaking, is a part of our implementation of ANN methods.

Reply:
Therefore, I think, if we want to move ann_kmeans_balanced into the cluster namespace, we should make the mapping public (at least non-dependent on ANN) as well.

Reply:
I think a first step is to change `copyRows` to take an iterator and then do the mapping with `cub::TransformInputIterator`. I've applied similar techniques to use raft prims in a WIP branch.

For the copy of the norm I'll replace it right now with `thrust::gather` though.

Reply:
You can go this way, sure, but there still is the logical issue of how we expect a dataset should be transformed within k-means algorithm (which right now is hardcoded in the mapping type).

Reply:
Yes, but in that regard, it matters little whether that hardcoded mapping type is given through an iterator or directly in the kernel such as in `copy_selected_kernel`, I'd even argue I prefer the former.

Reply:
Ah, yes, I totally agree (just make sure we don't lose much in terms of performance). I meant to explain issues with the other topic discussed above - moving ann_kmeans_balanced into the cluster namespace, whereas it currently contains ANN-implementation-specific logic for transforming the data.

Reply:
@Nyrio trying to use raft primitives where at all possible, would raft::matrix::gather work in place of thrust::gather? 

Also, just FYI, we are in the process of migrating the RAFT APIs to accept mdspan everywhere to make them consistent across the board, cleaner, easier to use. Maybe we could also use raft::span (which can construct an interator)?

Reply:
@cjnolet `matrix::gather` will be very inefficient for `D=1`:

```cuda
for (int i = threadIdx.x; i < D; i += TPB) {
  out[outRowStart + i] = in[inRowStart + i];
}
```
We could have an array/vector gather but it would essentially be a simple wrapper around thrust.

Reply:
Note that I used `thrust::gather` only for the norm, we could however use `matrix::gather` to select rows in the dataset (in another PR). Its support for iterators is convenient, I could use that to do the int-float mapping.

Don't we have some redundancy between `copyRows` and `gather`? The latter seems to be a generalization of the former, am I right?